### PR TITLE
Implement <source_location>

### PR DESCRIPTION
**Pull request by SuperWig**
Description
===========
**This PR is blocked until issues with the compiler builtins are solved** (as well as `consteval`?)

So after now being able to run the tests there's some more potential oddities I found. Thought may as well open a draft PR as it might be helpful for compiler devs.

So following what I found in #54. 

* Tests 17 and 18 fail the column number asserts *except* when `current()` is used as a default argument (i.e. 1st `argument` test, `sloc`, and `lambda` pass)
* ```
   source_location::current();
   ^
   source_location::current();
                    ^
   ```
   When used as a lambda capture the column number is the first otherwise the second.

### Issues to resolve
- [ ] `consteval` `source_location::current()`
- [ ] `constexpr` `__builtin_FILE()` and `__builtin_FUNCTION()`
- [ ] Consistent column number location (`lambda_test()`)
- [ ] Column number when not a default argument under tests 17 and 18 
- [x] Column number with tabs
- [x] `sub_member_test()`
- [x] ~~NTTP test~~
----

I'm unsure how to test the tab column issue as that would mean this file fails validation, right? Thought maybe put it in a different file with a file extension that could be skipped e.g. `test.tab`?

Is using `ends_with` sufficient for the file name tests?

Checklist
=========

Be sure you've read README.md and understand the scope of this repo.

If you're unsure about a box, leave it unchecked. A maintainer will help you.

- [x] Identifiers in product code changes are properly `_Ugly` as per
  https://eel.is/c++draft/lex.name#3.1 or there are no product code changes.
- [ ] The STL builds successfully and all tests have passed (must be manually
  verified by an STL maintainer before automated testing is enabled on GitHub,
  leave this unchecked for initial submission).
- [x] These changes introduce no known ABI breaks (adding members, renaming
  members, adding virtual functions, changing whether a type is an aggregate
  or trivially copyable, etc.).
- [x] These changes were written from scratch using only this repository,
  the C++ Working Draft (including any cited standards), other WG21 papers
  (excluding reference implementations outside of proposed standard wording),
  and LWG issues as reference material. If they were derived from a project
  that's already listed in NOTICE.txt, that's fine, but please mention it.
  If they were derived from any other project (including Boost and libc++,
  which are not yet listed in NOTICE.txt), you *must* mention it here,
  so we can determine whether the license is compatible and what else needs
  to be done.

### CODE REVIEW

File: tests/std/tests/P1208R6_source_location/test.cpp
Comment (SuperWig) [reply tests/std/tests/P1208R6_source_location/test.cpp]: Good point.

File: tests/std/tests/P1208R6_source_location/test.cpp
Comment (SuperWig) [reply tests/std/tests/P1208R6_source_location/test.cpp]: ...What an obvious solution to the problem. 😅

File: tests/std/tests/P1208R6_source_location/test.cpp
Comment (AdamBucior) [inline tests/std/tests/P1208R6_source_location/test.cpp]: These hard coded line values may get invalid pretty quickly while editing this file. I would suggest doing something like that, though I leave it up to you.
```suggestion
    argument_test(__LINE__, 5);
    auto loc = std::source_location::current();
    argument_test(__LINE__ - 1, 38, loc);
```

**Comment by StephanTLavavej** [comment]
Microsoft-internal VSO-778944 tracks `consteval` support in the compiler, currently being implemented by Jennifer Yao.

**Comment by xiangfan-ms** [comment]
Hi,
    The compiler intrinsics expose the internal source location in the compiler. The implementation doesn't try to change the existing source location. If you find some source locations confusing, we can take a closer look and see whether we can adjust them (though it often has implications on other things like diagnosis, debugging behavior, static analysis, etc.).

    For some of the issues you raised:
1. The source code is stored as UTF-8 internally. The column is the number of bytes between the current character and the first character at the beginning of the line. So tab is counted as 1 and a Japanese character may be counted as 3.
You can validate the column using /diagnostics:column
2. The source location of a qualified-id points to the start of the last component.
3. The ''s2 x;" example involves compiler generated function (ctor of s2). Its body reuses whatever the source location the last parsing left, so in general it isn't something reliable. You can validate using the debugger :-).

**Comment by SuperWig** [comment]
1. Thanks for clarifying  :)
2. I figured this would be special cased or something as it would be impossible to get a column number less than 18 with how it is now. Clang has it working the way I was expecting
https://godbolt.org/z/NAhDEW
3. Ah, so this should work how I was expecting? (it does with Clang)
   ```cpp
   struct s2 {
       s x = source_location::current();
   };
   ```

Note: I'm not arguing one way or the other for what the column numbers are supposed to be, just explaining my reasons for my expectations.

**Comment by StephanTLavavej** [comment]
Removing the `blocked` tag - while MSVC's frontend C1XX is still working on `consteval`, Clang supports it, and we can ship features active for Clang only at first (as we have done so in the past with CTAD).

**Comment by SuperWig** [comment]
Should I remove the test from the `test.lst`? I suppose the tab test isn't needed after that clarification, so that's not needed either, right?

**Comment by StephanTLavavej** [comment]
For a test that works for some compilers but not others, I think the current best way is to `#ifdef` within the test itself.

For tab characters, we have the ability to exempt files from the "no tabs anywhere" validation (see `tabby_filenames` in validate.cpp :cat2: ), but I think that the STL's tests should validate basic functionality of `source_location` - the behavior in corner cases like tab characters should be left up to the compiler. Other maintainers may feel differently about "end-to-end" testing here.

**Comment by SuperWig** [comment]
Had to change the header file name test as clang expects
```
tests\std\tests\P1208R6_source_location/header.h
```
while MSVC would expect
```
tests\std\tests\P1208R6_source_location\header.h
```
if I'm not mistaken. Or would it be prefered to do an `#ifdef` for that?

**Comment by StephanTLavavej** [comment]
I think what you did is fine for now, but I'd recommend filing a bug against Clang, reporting their behavior divergence from MSVC (I think they're the "strange" ones since they're not consistently using Windows' preferred separators in the path).

**Comment by SuperWig** [comment]
Not being all `\` or `/` was definitely strange when trying to figure out why that test was failing...

**Comment by SuperWig** [comment]
Should I turn the OP into an issue so I can add transition comments everywhere?

File: stl/inc/__msvc_all_public_headers.hpp
Comment (CaseyCarter) [inline stl/inc/__msvc_all_public_headers.hpp:105]: In this header conditional inclusion is only for things that will `#error` and blow up a compile - you should be fine to include `<source_location>` unconditionally.

File: tests/std/tests/P1208R6_source_location/test.cpp
Comment (CaseyCarter) [inline tests/std/tests/P1208R6_source_location/test.cpp:136]: Missing comment here.
```suggestion
#endif // __clang__
```

File: tests/std/tests/P1208R6_source_location/test.cpp
Comment (CaseyCarter) [inline tests/std/tests/P1208R6_source_location/test.cpp:134]: We typically describe both the above and below cases in an `#else` so a reader need not scroll back and find the corresponding `#if`:
```suggestion
#else // ^^^ Clang / not Clang vvv
```

**Review by CaseyCarter — DISMISSED** [review]

File: stl/inc/__msvc_all_public_headers.hpp
Comment (SuperWig) [reply stl/inc/__msvc_all_public_headers.hpp:105]: I thought that but it didn't work as it has no idea what `consteval` was. I have changed that to use the feature test macro, just haven't pushed it.

Thought it was strange that it needed that though...

File: tests/std/tests/P1208R6_source_location/test.cpp
Comment (SuperWig) [reply tests/std/tests/P1208R6_source_location/test.cpp:21]: Because I'm so used to giving them names I completely forgot you can just not name then🤦‍♂️...

**Comment by CaseyCarter** [comment]
> Should I turn the OP into an issue so I can add transition comments everywhere?

Yes, an issue to enable the feature for MSVC when the compiler support arrives would help us track.

File: stl/inc/__msvc_all_public_headers.hpp
Comment (CaseyCarter) [reply stl/inc/__msvc_all_public_headers.hpp:105]: Oh, I see, the conditional in `<source_location>` is broken.

File: stl/inc/source_location
Comment (CaseyCarter) [inline stl/inc/source_location:11]:
```suggestion
#if !_HAS_CXX20 || !defined(__clang__)
```

Comment (SuperWig) [reply stl/inc/source_location:11]: Haha! Early morning me thinking I'm fixing something https://github.com/microsoft/STL/pull/664/commits/d3f4c9ec81591e73a57a8e613198acfdaac1066a

Comment (CaseyCarter) [reply stl/inc/source_location:11]: From some quick-and-dirty checks, it looks like MSVC supports the builtins now, and is only lacking `consteval`. So ideally this would be `!_HAS_CXX20 || !defined(__cpp_consteval)`, except that Clang doesn't define `__cpp_consteval` despite implementing `consteval` (the feature-test macro was added to the C++20 working draft fairly late). So let's go with:
```suggestion
#if !_HAS_CXX20 || !(defined(__cpp_consteval) || defined(__clang__)) // TRANSITION, LLVM-45745
```
where LLVM-45745 is a bug nicely asking the clang devs to define the macro.

File: tests/std/tests/P1208R6_source_location/test.cpp
Comment (SuperWig) [reply tests/std/tests/P1208R6_source_location/test.cpp:136]: I thought that was a #351 issue. Though I'm now remembering they're always applied to the `#endif`, right?

File: tests/std/tests/P1208R6_source_location/test.cpp
Comment (SuperWig) [reply tests/std/tests/P1208R6_source_location/test.cpp:34]: As every variable can be `constexpr` and I have one use of a local `constexpr` variable, should that be `const` or should everything be `constexpr`?

**Comment by SuperWig** [comment]
I'll create the issue along with further review comments (like `unsigned int`?) in the morning.

File: stl/inc/source_location
Comment (CaseyCarter) [inline stl/inc/source_location:11]: We haven't changed this for hours now!

A Clang dev responded to LLVM-4575 to say that they intentionally do not define `__cpp_consteval` because the support for immediate functions in Clang is incomplete and not ready for production use. We should, therefore, not use it:
```suggestion
#if !_HAS_CXX20 || !defined(__cpp_consteval)
```

Comment (CaseyCarter) [reply stl/inc/source_location:11]: Yes, I think that makes this header effectively dead code at the moment. I suppose we should either bring back the blocked tag, or remove the `defined(__cpp_consteval)` guard and annotate `source_location::current` with `_CONSTEVAL` which allows it to be `constexpr` when the compiler doesn't do `consteval`. Since I suspect the latter would break this feature, I'll add the tag.

Comment (SuperWig) [reply stl/inc/source_location:11]: > We haven't changed this source line for hours now!

Sincerest apologies. I've made 3 commits to make up for that. I also thought to just use `__cpp_consteval` as that I wouldn't have thought that to be defined unconditionally. If that's an incorrect assumption we get to modify that line again! 🎉

Comment (CaseyCarter) [reply stl/inc/source_location]: > I also thought to just use `__cpp_consteval` as I wouldn't have thought that to be defined unconditionally

I left the `_HAS_CXX20` there mostly out of paranoia that some compiler would accidentally define the feature-test macro in a sub-C++20 language mode - not likely, but it has happened before. It's probably not worth the trouble to put it back: we can change this line later if we need to. I think we've established that we can change it over and over whenever we like ;)

Comment (SuperWig) [reply stl/inc/source_location]: Interesting... 

Well the main reason I opted for dropping `_HAS_CXX20` was to not have a super long `#else` comment or think of a shorter way to phrase it. 
Seeing as it's going to become just `_HAS_CXX20` eventually (I assume) it seemed like a correct solution that took the minimalist amount of effort.

**Comment by StephanTLavavej** [comment]
Jennifer's Microsoft-internal MSVC-PR-255854 implementing `consteval` is out for review, which means it might be merged for VS 2019 16.8 (no promises yet).

**Comment by StephanTLavavej** [comment]
`consteval` will be available in VS 2019 16.8 Preview 2 (unless a meteor hits) *except* that the feature-test macro won't be defined because there's one remaining part (involving constant-folding immediate invocations). The partial feature may be sufficient to allow this PR to make further progress, so we can investigate at that time. Preview 3 is still open for feature work (no promises or ETA yet).

**Comment by SuperWig** [comment]
> consteval will be available in VS 2019 16.8 Preview 2 (unless a meteor hits) 

So we're still good if it's the apocalypse?

Just realised I fogot to push a commit...

**Comment by SuperWig** [comment]
> there's one remaining part (involving constant-folding immediate invocations)

@StephanTLavavej  Would that be the reason I'm getting this in places where it's being used as a default argument?
```
'std::source_location::current': call to immediate function is not a constant expression
```

**Comment by StephanTLavavej** [comment]
VS 2019 16.8 Preview 2 is now available and installed on the CI machines; it may be possible to make further progress on this PR.

**Comment by SuperWig** [comment]
> I wonder if passing it through integral_constant would help.

I'm not exactly sure what you're suggesting to pass through `integral_constant` 😕 .

>  especially if it's inconsistent across compilers

Definitely the more important aspect. Users can add/subtract as they would like.
So I guess I'll need to create a macro for testing columns then 😞.

**Comment by StephanTLavavej** [comment]
To force constant evaluation, I was thinking that `integral_constant<int, function_call()>::value` might help - but I haven't looked into this specific error, so it might be totally useless.

**Comment by SuperWig** [comment]
Ah so you meant the function call. That doesn't help with the structs which doesn't allow 
```cpp
struct s {
    source_location l = source_location::current();
};
```
only
```cpp
struct s {
    constexpr s(source_location x = source_location::current()) : l{x} {}
    source_location l;
};
```

**Comment by StephanTLavavej** [comment]
I just checked, and unfortunately `__cpp_consteval` will not be defined in 16.8 Preview 3 (releasing in the near future), so I suspect that it won't contain a fix for this issue either.

**Comment by SuperWig** [comment]
I suppose this should be labelled blocked again, then? ☹️

**Comment by StephanTLavavej** [comment]
Yeah, looks like it (because Clang still doesn't support `consteval` either, even in Clang 11). I'll add the label again.

**Comment by StephanTLavavej** [comment]
Fix won't be available in 16.9 Preview 2; updating label to Preview 3 (hopefully).

**Comment by StephanTLavavej** [comment]
The latest word from the compiler team is that the fix is expected for 16.9 Preview 4.

**Comment by CaseyCarter** [comment]
The latest word from the compiler team is that the fix is expected for 16.9 Preview 5.

**Comment by StephanTLavavej** [comment]
A compiler PR is now being reviewed, which would be in time for 16.10 Preview 2. (It *may* be possible to port to Preview 1 if the stars align.)

**Comment by SuperWig** [comment]
Does the fact that this comes up with IntelliSense going to be an issue at all? (someone might want to pass something in? 🤷‍♂️)

image

I wouldn't *think* it would but, you never know...

File: stl/inc/source_location
Comment (MichaelRizkalla) [inline stl/inc/source_location:28]:
```suggestion
    _NODISCARD static consteval source_location current() noexcept {
```
I think this function should not take any parameters.

File: stl/inc/source_location
Comment (MichaelRizkalla) [inline stl/inc/source_location:30]:
```suggestion
        _Result._Line     = __builtin_LINE();
```

File: stl/inc/source_location
Comment (MichaelRizkalla) [inline stl/inc/source_location:31]:
```suggestion
        _Result._Column   = __builtin_COLUMN();
```

File: stl/inc/source_location
Comment (MichaelRizkalla) [inline stl/inc/source_location:32]:
```suggestion
        _Result._File     = __builtin_FILE();
```

File: stl/inc/source_location
Comment (MichaelRizkalla) [inline stl/inc/source_location:33]:
```suggestion
        _Result._Function = __builtin_FUNCTION();
```

**Review by MichaelRizkalla — CHANGES_REQUESTED** [review]
I have noticed that both [P1208R6 ](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2019/p1208r6.pdf) and [N4878](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2020/n4878.pdf) are declaring `source_location::current()` without any arguments. So I *think* the following changes should apply.

**Comment by ZaldronGG** [comment]
> Edit: It seems I was wrong, there's no way around it 🐱 unless there's an intermediate private struct to be used in the constructor

FWIW, this is almost identical to how the C# team handled [caller information](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/attributes/caller-information) (using attributes and bogus-defaulted arguments instead of builtins).

**Comment by StephanTLavavej** [comment]
WG21-N4878 [member.functions]/2: "For a non-virtual member function described in the C++ standard library, an implementation may declare a different set of member function signatures, provided that any call to the member function that would select an overload from the set of declarations described in this document behaves as if that overload were selected. [Note 1 : For instance, an implementation can add parameters with default values, or replace a member function with default arguments with two or more member functions with equivalent behavior, or add additional signatures for a member function name. — end note]"

This permits the current implementation technique. I don't think we need to worry too much about people writing non-Standard code here.

**Comment by DaemonSnake** [comment]
Hi,
I was wondering what is the expected behavior of calling std::source_location::current().function_name() from within a lambda.
GCC and CLANG's experimental version and MSVC __builtin_FUNCTION() produce the "operator()" string but the non-experimental version of GCC (from trunk) produce "main()::<lambda()>  (you can check it out here: https://godbolt.org/z/6dMTbd).
Thanks in advance

**Comment by BurningEnlightenment** [comment]
> I was wondering what is the expected behavior of calling std::source_location::current().function_name() from within a lambda.
GCC and CLANG's experimental version and MSVC __builtin_FUNCTION() produce the "operator()" string but the non-experimental version of GCC (from trunk) produce "main()::<lambda()>

You can expect an _implementation defined_ string as per [`[support.srcloc.cons]`](https://eel.is/c++draft/support.srcloc.class#support.srcloc.cons-1.1). Which specifies the behaviour of `source_location::current()`:
> When invoked by a function call [...] returns a `source_­location` with an implementation-defined value. [...]

> A name of the current function such as in `__func__` `([dcl.fct.def.general])` if any, an empty string otherwise.

However, their `__PRETTY_FUNCTION__` / `__FUNCSIG__` format diverges quite a bit, so I would be surprised if MSVC, GCC and Clang aligned on a common format for `function_name()`.

In any case I find it a bit saddening that the currently available implementation of `__builting_FUNCTION()` delivers less useful information than `__FUNCSIG__`, but I still hope for future improvements there.

**Comment by StephanTLavavej** [comment]
I've pushed a batch of changes:

* Merge `main`, manually resolving `yvals_core.h` to add the feature-test macro in sorted order.
* Add `source_location` to the current set of ["hack these files"](https://github.com/microsoft/STL/wiki/Files-To-Edit-When-Adding-Or-Removing-Files). Haven't attempted to add actual code to the Standard Library Header Units test yet (commented with a FIXME).
* As `<source_location>` is a "core" header, I've moved it to the appropriate section in `__msvc_all_public_headers.hpp`. I've also updated the wiki to note these sections (as I noticed it only by chance).
* Sort the addition in the feature-test macro test.
* Get the test to compile with VS 2019 16.10 Preview 2; this requires changing a bunch of expected column numbers (possibly bugs) and a couple of expected function names only at compile-time (definitely bugs), all noted with FIXMEs.

There are also at least 2 issues with EDG (`constexpr` errors and unexpected column numbers); I didn't investigate those closely, but they will need to be analyzed and reported.

**Comment by SuperWig** [comment]
> There are also at least 2 issues with EDG (constexpr errors and unexpected column numbers);

EDG is expecting the column numbers to be after the opening paren. i.e.
```
source_location::current( );
                         ^ here
```
After doing an `#ifdef` for those cases, the remaining issues were with default arguments/in-class initialisation
```
"E:\Projects\Repos\MicrosoftSTL\tests\std\tests\P1208R6_source_location\test.cp
          p", line 26: error: call to consteval function
          "std::source_location::current" did not produce a valid constant
          expression
      constexpr s(const source_location x = source_location::current()) : loc(x) {}
                                            ^

"E:\Projects\Repos\MicrosoftSTL\tests\std\tests\P1208R6_source_location\test.cp
          p", line 28: error: call to consteval function
          "std::source_location::current" did not produce a valid constant
          expression
      source_location loc = source_location::current();
                            ^

"E:\Projects\Repos\MicrosoftSTL\tests\std\tests\P1208R6_source_location\test.cp
          p", line 33: error: call to consteval function
          "std::source_location::current" did not produce a valid constant
          expression
      constexpr s2(const source_location l = source_location::current()) : x{l} {}
                                             ^

"E:\Projects\Repos\MicrosoftSTL\tests\std\tests\P1208R6_source_location\test.cp
          p", line 35: error: call to consteval function
          "std::source_location::current" did not produce a valid constant
          expression
      s x = source_location::current();
            ^

"E:\Projects\Repos\MicrosoftSTL\tests\std\tests\P1208R6_source_location\test.cp
          p", line 60: error: call to consteval function
          "std::source_location::current" did not produce a valid constant
          expression
      const unsigned int line, const unsigned int column, const source_location x = source_location::current()) {
```

**Comment by StephanTLavavej** [comment]
Thanks, I've internally reported VSO-1285779 "EDG rejects `source_location` intrinsics within `consteval` functions" and (for C1XX) VSO-1285783 "C++20 `source_location` intrinsic `__builtin_FUNCTION()` behaves differently when `is_constant_evaluated`".

**Comment by SuperWig** [comment]
Unless clang plans on changing their column number, shouldn't there be `#ifdef`s for clang?

**Comment by StephanTLavavej** [comment]
We'll need to add ifdefs for Clang and EDG when we can activate them.

**Review by StephanTLavavej — APPROVED** [review]

**Comment by SuperWig** [comment]
Actually, since Clang is going to be self activating would it not make sense to pre-emptively have a function like this to compare against instead?
```cpp
constexpr int get_c(int n)
{
#ifdef __clang__
    return n - 17;
#elif defined(__EDG__)
    return n + 8;
#else
    return n;
#endif
}
```
That way its less work for the PR that updates Clang to a version that defines `__cpp_consteval`

**Comment by StephanTLavavej** [comment]
I would prefer to defer any work for Clang until the toolset update that activates Clang - because I want to land this *very* soon, to meet the 16.10 Preview 2 deadline, and I just finished validating this against the internal and public builds, so resetting testing would add additional headache and risk. Your suggested function sounds like a great idea and I'll try to remember it (since I expect to perform that toolset update). Note that `consteval` [still isn't implemented](https://clang.llvm.org/cxx_status.html) in Clang 12 so it'll be a while.

Of the changes that I recently pushed, the notable ones are:

* When testing `file_name` `ends_with`, use only `test.cpp` or `header.h`, not the relative path with the test directory. This is because (AFAIK) the internal "Contest" test harness copies files to temporary directories before compiling them. As a result, LLVM-45731 is no longer marked as TRANSITION (they can continue to investigate it, of course).
* `/analyze` expects slightly different column numbers. This might be a bug - similarly for Standard Library Header Units returning `column() == 1`. **For all column number behavior, bugs can be reported to the various compilers, but I don't think they should block this checkin.**
* `include_each_header_alone_matrix.lst` is gaining an MSVC configuration. We initially tested only Clang because it provided a highly conformant mode with two-phase name lookup. Now that MSVC has `/permissive-` strict mode, and we are increasingly implementing features like `<source_location>` that are active for one compiler earlier than another, having coverage for both codegen compilers is increasingly important. Fortunately, this revealed no issues.
* We should use `_NODISCARD_CTOR` for constructors to be consistent, although I believe it makes no difference for C++20-only code.

I've mirrored this to internal MSVC-PR-307468 along with the necessary "hack these files" changes to add `<source_location>` to the installer and so forth.

Please note: Further changes are still possible (e.g. if you notice something, or if final review leads to more feedback that should be addressed before merging), but they must be replicated to the mirror PR in order to avoid codebase divergence.

**Comment by SuperWig** [comment]
Thanks for the super detailed answer as always, Stephan.

File: tests/std/tests/VSO_0157762_feature_test_macros/test.compile.pass.cpp
Comment (AnjuDel) [inline tests/std/tests/VSO_0157762_feature_test_macros/test.compile.pass.cpp:1361]: rewriting this like
 #else //^^^__cpp_lib_source_location / __cpp_lib_source_location vvv 
 will help this make this more readable.
 ditto on other elses elif and endif

**Review by AnjuDel — CHANGES_REQUESTED** [review]

File: tests/std/tests/VSO_0157762_feature_test_macros/test.compile.pass.cpp
Comment (StephanTLavavej) [reply tests/std/tests/VSO_0157762_feature_test_macros/test.compile.pass.cpp:1361]: The lack of comments in the feature-test macro test is intentionally different from the rest of the codebase; basically these blocks of preprocessor logic are dense and compact, so the comments don’t help clarity they way they do in normal code.

File: tests/std/tests/P1208R6_source_location/test.cpp
Comment (StephanTLavavej) [reply tests/std/tests/P1208R6_source_location/test.cpp]: I think the current state is good. We strongly prefer to add `const` when possible (unless a test needs non-`const` which is fairly common) but aren't nearly as concerned with making everything as `constexpr` as possible (e.g. `const int i = 1729;` is common). The dual-mode runtime/compile-time testing ensures that everything is usable in constant expressions.

File: tests/std/tests/P1208R6_source_location/test.cpp
Comment (StephanTLavavej) [reply tests/std/tests/P1208R6_source_location/test.cpp]: The `#endif` now has a comment. Our current semi-consistent convention is to either omit the `#else` and `#endif` comments together (when they span only 1-2 lines and aren't near other preprocessor logic), or have both. I consider any cases where only one has a comment to be undesirable.

File: stl/inc/source_location
Comment (StephanTLavavej) [reply stl/inc/source_location]: I think the current state (guarding with only `__cpp_consteval`) is fine. For comparison, `<concepts>` is guarded with only `__cpp_lib_concepts`. (I suppose we could change the guard to `__cpp_lib_source_location` to truly centralize everything, but that would be a minor cleanup at best.)

**Comment by StephanTLavavej** [comment]
Thanks for implementing this feature far in advance of the compilers being ready! :joy_cat: This will be available in VS 2019 16.10 Preview 2 along with MSVC's support for `consteval`. :rocket: :smiley_cat: :world_map: :dart:

File: stl/inc/source_location
Comment (robert-andrzejuk) [inline stl/inc/source_location]: @StephanTLavavej [INDENTATION STYLE RANT] Theres some ugly indentation here. I will suggest a lecture by Kavlin Henney: ITT 2016 - Kevlin Henney - Seven Ineffective Coding Habits of Many Programmers https://www.youtube.com/watch?t=1263&v=ZsHMHukIlJY [/INDENTATION STYLE RANT] 

The below formatting, I believe I get by using clang-format parameters (https://clang.llvm.org/docs/ClangFormatStyleOptions.html):
_AllowAllConstructorInitializersOnNextLine: false
AllowAllParametersOfDeclarationOnNextLine: false
BinPackArguments: false
BinPackParameters: false
AlignConsecutiveAssignments: AcrossComments
AlignConsecutiveDeclarations: AcrossComments_

```
static constexpr source_location current(uint_least32_t _InLine     = __builtin_LINE(),
                                         uint_least32_t _InColumn   = __builtin_COLUMN(),
                                         const char*    _InFile     = __builtin_FILE(),
                                         const char*    _InFunction = __builtin_FUNCTION()) noexcept
{
    source_location _Result;
    _Result._Line     = _InLine;
    _Result._Column   = _InColumn;
    _Result._File     = _InFile;
    _Result._Function = _InFunction;
    return _Result;
}
```

As style is a contentious issue, please feel free to ignore this rant.

**Pull request merged**
